dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java

```diff
@@ -76,13 +87,118 @@
       ConstantNameAndType constantNameAndType = (ConstantNameAndType) constantAtNameAndTypeIndex;
       String methodName = constantNameAndType.getName(constantPool);
       String descriptor = constantNameAndType.getSignature(constantPool);
-      FullyQualifiedMethodSignature methodref = new FullyQualifiedMethodSignature(className,
-          methodName, descriptor);
+      FullyQualifiedMethodSignature methodref =
+          new FullyQualifiedMethodSignature(classNameInMethodReference, methodName, descriptor);
       methodReferences.add(methodref);
     }
     return methodReferences;
   }
 
+  /**
+   * Lists external method references for a class. The returned list does not include method
+   * references that point to other classes defined in the same jar file as the class.
+   *
+   * @param className class name to list its method references. The class must be available through
+   *     the class loader and BCEL repository.
+   * @param jarFileToClasses mapping of jar file paths to classes. This helps to distinguish whether
+   *     the class in a method reference is in the same jar file or not.
+   * @param classLoader class loader to locate the jar file for the class
+   * @param repository BCEL repository to list method references for the class
+   * @return list of external method references from the class
+   */
+  static List<FullyQualifiedMethodSignature> listExternalMethodReferences(
+      String className,
+      SetMultimap<Path, String> jarFileToClasses,
+      ClassLoader classLoader,
+      SyntheticRepository repository) {
+    try {
+      JavaClass javaClass = repository.loadClass(className);
+      Class clazz = classLoader.loadClass(className);
+      CodeSource codeSource = clazz.getProtectionDomain().getCodeSource();
+      if (codeSource == null) {
+        // Code in bootstrap class loader (e.g., javax) does not have source
+        return Collections.emptyList();
+      }
+      Path jarPathForTheClass = Paths.get(codeSource.getLocation().toURI());
+      Set<String> classesDefinedInSameJar = jarFileToClasses.get(jarPathForTheClass);
+      List<FullyQualifiedMethodSignature> nextMethodReferences = listMethodReferences(javaClass);
+      List<FullyQualifiedMethodSignature> nextExternalMethodReferences = new ArrayList<>();
+      List<String> referencedInternalClasses = new ArrayList<>();
+      for (FullyQualifiedMethodSignature methodReference : nextMethodReferences) {
```

Comment:
This looks like the same loop as below, in `findExternalMethodReferencesByUsageGraph()`. Instead of copying the loop here in order to get the initial `referencedInternalClasses`, could you just call that method with a list containing just `className`?

That is, skip this loop entirely. Right after initializing `classesDefinedInSameJar` above, do this:

```
return findExternalMethodReferencesByUsageGraph(
    new ArrayList<>(className), repository, classesDefinedInSameJar);
```